src/compiler/crystal/syntax/parser.cr

```diff
@@ -2451,6 +2459,40 @@ module Crystal
       Case.new(cond, whens, a_else)
     end
 
+    # Add an expression to all when expressions and error on duplicates
+    def add_when_exp(when_exps, exp)
+      return unless when_exp_constant?(exp)
+
+      if when_exps.includes?(exp)
+        raise "duplicate when #{exp} in case", exp.location.not_nil!
+      end
+
+      when_exps << exp
+    end
+
+    # Only error on constant values, because calls might have side-effects:
+    # a first call might return one value and not match the case
+    # value, but the second same call returns something different
+    # and matches it.
+    def when_exp_constant?(exp)
+      case exp
+      when NilLiteral, BoolLiteral, CharLiteral, NumberLiteral,
+           StringLiteral, SymbolLiteral, Path
+        true
+      when ArrayLiteral
+        exp.elements.all? { |e| when_exp_constant?(e) }
+      when TupleLiteral
+        exp.elements.all? { |e| when_exp_constant?(e) }
+      when RegexLiteral
+        when_exp_constant?(exp.value)
+      when RangeLiteral
+        when_exp_constant?(exp.from) &&
+          when_exp_constant?(exp.to)
```

Comment:
Looks like here we need to test `{from, to}` as single expression.

Reply:
What do you mean? Here I just check that both ends are "constants" as well. The check for duplicates is in the `includes?` call in `add_when_exp`.

Reply:
```crystal
when 5..10
when 6
when 7..11
```

Reply:
It works for me. `5..10`, `6` and `7..11` are all different AST nodes, so the duplicate error won't trigger.

Reply:
But they should trigger error.

Anyway when you enables these intersections, you must to check `{from, to}` as single value instead of two separated.

```crystal
when 2..5
when 3..5
when 2..4
```

will fail now afaiu.

Reply:
Oh, I see what you mean. This change is just about detecting duplicates, what you ask is much more complex. Maybe it can be done for simple ranges like this, but we can do it in a separate PR.

Did you try that code? It compiles fine. I don't understand why you say it will fail now.

Reply:
I was wrong, you are right here. But yes, intersection checks should be applied, but it's subject of another PR.